src/sas/sascalc/fit/fitstate.py

```diff
@@ -0,0 +1,407 @@
+    dists.update(state.disp_obj_dict)
+    for par_name, dist_name in state.disp_obj_dict.items():
+        dispersion = POLYDISPERSITY_MODELS[dist_name]()
+        if dist_name == "array":
+            dispersion.set_weights(state.values[par_name], state.weights[par_name])
+        base_par = par_name.replace('.width', '')
+        model.set_dispersion(base_par, dispersion)
+
+    # Put parameter values and ranges into the model
+    fitted = []
+    for par_tuple in state.parameters + state.fixed_param + state.fittable_param:
+        par = SasviewParameter(*par_tuple)
+        if par.name not in state.weights:
+            # Don't try to set parameter values for array distributions
+            # TODO: keep weights filename in the array distribution object
+            model.setParam(par.name, parse_optional_float(par.value))
+        if par.fitted:
+            fitted.append(par.name)
+        if par.name in model.details:
+            lower = parse_optional_float(par.lower[1])
+            upper = parse_optional_float(par.upper[1])
+            model.details[par.name] = [par.units, lower, upper]
+    #print("pars", model.params)
+    #print("limits", model.details)
+    #print("fitted", fitted)
+
+    # Set the resolution
+    data = copy.deepcopy(state.data)
+    if state.disable_smearer:
+        smearer = None
+    elif state.enable_smearer:
+        smearer = smear_selection(data, model)
+    elif state.pinhole_smearer:
+        # see sasgui/perspectives/fitting/basepage.py: reset_page_helper
+        dx_percent = state.dx_percent
+        if state.dx_old:
+            dx_percent = 100*(state.dx_percent / data.x[0])
+        # see sasgui/perspectives/fitting/fitpage.py: _set_pinhole_smear
+        percent = dx_percent / 100.
+        if state.enable2D:
+            # smear_type is Pinhole2D.
+            # dqx_data, dqy_data initialized to 0
+            data.dqx_data = percent * data.qx_data
+            data.dqy_data = percent * data.qy_data
+        else:
+            data.dx = percent * data.x
+            data.dxl = data.dxw = None  # be sure it is not slit-smeared
+        smearer = smear_selection(data, model)
+    elif state.slit_smearer:
+        # see sasgui/perspectives/fitting/fitpage.py: _set_pinhole_smear
+        data_len = len(data.x)
+        data.dx = None
+        data.dxl = (state.dxl if state.dxl is not None else 0.) * np.ones(data_len)
+        data.dxw = (state.dxw if state.dxw is not None else 0.) * np.ones(data_len)
+        smearer = smear_selection(data, model)
+    else:
+        raise ValueError("expected resolution specification for fit")
+
+    # Set the data weighting (dI, sqrt(I), I, or uniform)
+    weight = get_data_weight(state)
+
+    # Make fit data object and set the data weights
+    # TODO: check 2D masked data
+    if state.enable2D:
+        fitdata = FitData2D(sas_data2d=data, data=data.data,
+                            err_data=data.err_data)
+        fitdata.err_data = weight
```

Comment:
Is fitdata.err_data = weight correct? There is also data.err_data and therefore I am asking

Reply:
This looks wrong.  Going back to the fitproblem.py code where I pulled this from, I think weight was being assigned to a copy of the original data object rather than the FitData2D object.  So, yes, it needs to be fixed and tested against a wider variety of cases (including masking, resolution, and weighting variants).